File: bindings/python/src/kriya/protocol.py
"""The agent-loop wire protocol the Python binding speaks to ``kriya-host`` over stdio.

Each line on the wire is a JSON object ``{"type": ..., "data": {...}}``. The host emits one object
per line on stdout (host -> app) and reads one per line on stdin (app -> host). The ``data`` fields
are **camelCase** on the wire (the Rust structs use ``#[serde(rename_all = "camelCase")]``), except
``done`` whose fields are already lowercase. These dataclasses use idiomatic snake_case and convert
at the edge via ``from_wire`` / ``to_wire``, so app code never juggles casing.

Mirrors kriya-core's ``protocol.ts`` and the Rust ``crates/kriya/src/protocol.rs`` / ``sidecar.rs``.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional


# -- host -> app -------------------------------------------------------------


@dataclass
class ActionRequest:
    """host -> app: run this action and reply with an :class:`ActionResultMsg` (same ``step_id``)."""

    step_id: str
    action_id: str
    params: Dict[str, Any]
    reasoning: str = ""

    @classmethod
    def from_wire(cls, d: Dict[str, Any]) -> "ActionRequest":
        return cls(
            step_id=d["stepId"],
            action_id=d["actionId"],
            params=d.get("params") or {},
            reasoning=d.get("reasoning", ""),
        )


@dataclass
class ApprovalRequest:
    """host -> app: this guarded action needs a human's go-ahead. Reply with
    :class:`ApprovalResponse` (same ``step_id``)."""

    step_id: str
    action_id: str
    params: Dict[str, Any]
    reasoning: str = ""

    @classmethod
    def from_wire(cls, d: Dict[str, Any]) -> "ApprovalRequest":
        return cls(
            step_id=d["stepId"],
            action_id=d["actionId"],
            params=d.get("params") or {},
            reasoning=d.get("reasoning", ""),
        )


@dataclass
class AwaitStep:
    """host -> app: paused before a step (step-mode only). Reply with :class:`StepAdvance`."""

    gate_id: str
    step_number: int
    last_action_id: Optional[str] = None
    last_success: Optional[bool] = None

    @classmethod
    def from_wire(cls, d: Dict[str, Any]) -> "AwaitStep":
        return cls(
            gate_id=d["gateId"],
            step_number=d.get("stepNumber", 0),
            last_action_id=d.get("lastActionId"),
            last_success=d.get("lastSuccess"),
        )


@dataclass
class Done:
    """host -> app: the run finished. (Wire fields are already lowercase: ``summary``/``steps``.)"""

    summary: str
    steps: int

    @classmethod
    def from_wire(cls, d: Dict[str, Any]) -> "Done":
        return cls(summary=d.get("summary", ""), steps=d.get("steps", 0))


@dataclass
class LogEntry:
    """host -> app: inspector/governance telemetry (reasoning, decisions, errors)."""

    level: str
    message: str
    step_id: Optional[str] = None
    detail: Any = None

    @classmethod
    def from_wire(cls, d: Dict[str, Any]) -> "LogEntry":
        return cls(
            level=d.get("level", "info"),
            message=d.get("message", ""),
            step_id=d.get("stepId"),
            detail=d.get("detail"),
        )


@dataclass
class Episode:
    """One recorded action from the host's durable episodic memory (newest-first).

    ``params`` is a JSON-encoded string exactly as the host signed it (parse it if you need the
    object). Mirrors the Rust ``kriya::memory::Episode`` and kriya-core's ``Episode``.
    """

    ts_ms: int
    action_id: str
    params: str
    success: bool
    reasoning: str
    signature: str
    run_id: str
    goal: str

    @classmethod
    def from_wire(cls, d: Dict[str, Any]) -> "Episode":
        return cls(
            ts_ms=d.get("tsMs", 0),
            action_id=d.get("actionId", ""),
            params=d.get("params", ""),
            success=bool(d.get("success", False)),
            reasoning=d.get("reasoning", ""),
            signature=d.get("signature", ""),
            run_id=d.get("runId", ""),
            goal=d.get("goal", ""),
        )


# -- app -> host -------------------------------------------------------------


@dataclass
class StartRequest:
    """app -> host: begin an autonomous run."""

    goal: str
    state: Dict[str, Any]
    tools: List[Dict[str, Any]] = field(default_factory=list)
    resume: bool = False
    step_mode: bool = False
    agent_id: Optional[str] = None
    user_id: Optional[str] = None

    def to_wire(self) -> Dict[str, Any]:
        out: Dict[str, Any] = {
            "goal": self.goal,
            "state": self.state,
            "tools": self.tools,
            "resume": self.resume,
            "stepMode": self.step_mode,
        }
        if self.agent_id is not None:
            out["agentId"] = self.agent_id
        if self.user_id is not None:
            out["userId"] = self.user_id
        return out


@dataclass
class ActionResultMsg:
    """app -> host: the result of an action the host asked you to run, plus refreshed state."""

    step_id: str
    success: bool
    state: Dict[str, Any]
    data: Any = None
    error: Optional[str] = None

    def to_wire(self) -> Dict[str, Any]:
        out: Dict[str, Any] = {
            "stepId": self.step_id,
            "success": self.success,
            "state": self.state,
        }
        if self.data is not None:
            out["data"] = self.data
        if self.error is not None:
            out["error"] = self.error
        return out


@dataclass
class ApprovalResponse:
    """app -> host: a human's decision on a pending approval."""

    step_id: str
    approved: bool

    def to_wire(self) -> Dict[str, Any]:
        return {"stepId": self.step_id, "approved": self.approved}


@dataclass
class StepAdvance:
    """app -> host: advance (``proceed=True``) or stop (``False``) a step-mode run."""

    gate_id: str
    proceed: bool

    def to_wire(self) -> Dict[str, Any]:
        return {"gateId": self.gate_id, "proceed": self.proceed}


__all__ = [
    "ActionRequest",
    "ApprovalRequest",
    "AwaitStep",
    "Done",
    "LogEntry",
    "Episode",
    "StartRequest",
    "ActionResultMsg",
    "ApprovalResponse",
    "StepAdvance",
]

File: bindings/dotnet/Kriya/Protocol.cs
using System.Text.Json.Nodes;

namespace Kriya;

// The agent-loop wire protocol the .NET binding speaks to kriya-host over stdio. Each line is a JSON
// object {"type": ..., "data": {...}}. `data` fields are camelCase (the Rust structs use
// rename_all="camelCase"), except `done` whose fields are already lowercase. These types use
// idiomatic PascalCase and convert at the edge via FromWire / ToWire — mirrors protocol.ts / the
// Python binding.

internal static class W
{
    public static string Str(JsonObject d, string k) => d[k]?.GetValue<string>() ?? "";
    public static string? StrOrNull(JsonObject d, string k) => d[k]?.GetValue<string>();
    public static JsonObject Obj(JsonObject d, string k) => d[k] as JsonObject ?? new JsonObject();
    public static bool Bool(JsonObject d, string k) => d[k]?.GetValue<bool>() ?? false;
    public static bool? BoolOrNull(JsonObject d, string k) => d[k] is JsonValue v && v.TryGetValue<bool>(out var b) ? b : null;
    public static long Long(JsonObject d, string k) => d[k]?.GetValue<long>() ?? 0;
    public static int Int(JsonObject d, string k) => d[k]?.GetValue<int>() ?? 0;
}

// ── host → app ──────────────────────────────────────────────────────────────

/// <summary>host → app: run this action and reply with an <see cref="ActionResultMsg"/> (same StepId).</summary>
public sealed class ActionRequest
{
    public string StepId { get; init; } = "";
    public string ActionId { get; init; } = "";
    public JsonObject Params { get; init; } = new();
    public string Reasoning { get; init; } = "";

    public static ActionRequest FromWire(JsonObject d) => new()
    {
        StepId = W.Str(d, "stepId"),
        ActionId = W.Str(d, "actionId"),
        Params = W.Obj(d, "params"),
        Reasoning = W.Str(d, "reasoning"),
    };
}

/// <summary>host → app: a guarded action needs a human's go-ahead. Reply with <see cref="ApprovalResponse"/>.</summary>
public sealed class ApprovalRequest
{
    public string StepId { get; init; } = "";
    public string ActionId { get; init; } = "";
    public JsonObject Params { get; init; } = new();
    public string Reasoning { get; init; } = "";

    public static ApprovalRequest FromWire(JsonObject d) => new()
    {
        StepId = W.Str(d, "stepId"),
        ActionId = W.Str(d, "actionId"),
        Params = W.Obj(d, "params"),
        Reasoning = W.Str(d, "reasoning"),
    };
}

/// <summary>host → app: paused before a step (step-mode). Reply with <see cref="StepAdvance"/>.</summary>
public sealed class AwaitStep
{
    public string GateId { get; init; } = "";
    public int StepNumber { get; init; }
    public string? LastActionId { get; init; }
    public bool? LastSuccess { get; init; }

    public static AwaitStep FromWire(JsonObject d) => new()
    {
        GateId = W.Str(d, "gateId"),
        StepNumber = W.Int(d, "stepNumber"),
        LastActionId = W.StrOrNull(d, "lastActionId"),
        LastSuccess = W.BoolOrNull(d, "lastSuccess"),
    };
}

/// <summary>host → app: the run finished. (Wire fields are lowercase: summary/steps.)</summary>
public sealed class Done
{
    public string Summary { get; init; } = "";
    public int Steps { get; init; }

    public static Done FromWire(JsonObject d) => new() { Summary = W.Str(d, "summary"), Steps = W.Int(d, "steps") };
}

/// <summary>host → app: inspector/governance telemetry.</summary>
public sealed class LogEntry
{
    public string Level { get; init; } = "info";
    public string Message { get; init; } = "";
    public string? StepId { get; init; }

    public static LogEntry FromWire(JsonObject d) => new()
    {
        Level = d["level"]?.GetValue<string>() ?? "info",
        Message = W.Str(d, "message"),
        StepId = W.StrOrNull(d, "stepId"),
    };
}

/// <summary>One recorded action from the host's durable memory (newest-first). Params is the
/// JSON-encoded string exactly as signed. Mirrors kriya::memory::Episode.</summary>
public sealed class Episode
{
    public long TsMs { get; init; }
    public string ActionId { get; init; } = "";
    public string Params { get; init; } = "";
    public bool Success { get; init; }
    public string Reasoning { get; init; } = "";
    public string Signature { get; init; } = "";
    public string RunId { get; init; } = "";
    public string Goal { get; init; } = "";

    public static Episode FromWire(JsonObject d) => new()
    {
        TsMs = W.Long(d, "tsMs"),
        ActionId = W.Str(d, "actionId"),
        Params = W.Str(d, "params"),
        Success = W.Bool(d, "success"),
        Reasoning = W.Str(d, "reasoning"),
        Signature = W.Str(d, "signature"),
        RunId = W.Str(d, "runId"),
        Goal = W.Str(d, "goal"),
    };
}

// ── app → host ──────────────────────────────────────────────────────────────

/// <summary>app → host: begin an autonomous run.</summary>
public sealed class StartRequest
{
    public required string Goal { get; init; }
    public JsonObject State { get; init; } = new();
    public IReadOnlyList<JsonObject> Tools { get; init; } = System.Array.Empty<JsonObject>();
    public bool Resume { get; init; }
    public bool StepMode { get; init; }
    public string? AgentId { get; init; }
    public string? UserId { get; init; }

    public JsonObject ToWire()
    {
        var tools = new JsonArray();
        foreach (var t in Tools) tools.Add(t.DeepClone());
        var o = new JsonObject
        {
            ["goal"] = Goal,
            ["state"] = State.DeepClone(),
            ["tools"] = tools,
            ["resume"] = Resume,
            ["stepMode"] = StepMode,
        };
        if (AgentId is not null) o["agentId"] = AgentId;
        if (UserId is not null) o["userId"] = UserId;
        return o;
    }
}

/// <summary>app → host: the result of an action the host asked you to run, plus refreshed state.</summary>
public sealed class ActionResultMsg
{
    public required string StepId { get; init; }
    public required bool Success { get; init; }
    public JsonObject State { get; init; } = new();
    public object? Data { get; init; }
    public string? Error { get; init; }

    public JsonObject ToWire()
    {
        var o = new JsonObject
        {
            ["stepId"] = StepId,
            ["success"] = Success,
            ["state"] = State.DeepClone(),
        };
        if (Data is not null) o["data"] = System.Text.Json.JsonSerializer.SerializeToNode(Data);
        if (Error is not null) o["error"] = Error;
        return o;
    }
}

/// <summary>app → host: a human's decision on a pending approval.</summary>
public sealed class ApprovalResponse
{
    public required string StepId { get; init; }
    public required bool Approved { get; init; }
    public JsonObject ToWire() => new() { ["stepId"] = StepId, ["approved"] = Approved };
}

/// <summary>app → host: advance (true) or stop (false) a step-mode run.</summary>
public sealed class StepAdvance
{
    public required string GateId { get; init; }
    public required bool Proceed { get; init; }
    public JsonObject ToWire() => new() { ["gateId"] = GateId, ["proceed"] = Proceed };
}
